devicetypes/vision-raytseng/vision-4-in-1-motion-sensor.src/vision-4-in-1-motion-sensor.groovy

```diff
@@ -0,0 +1,495 @@
+	}
+	if (device?.currentValue("illuminance") == null) {
+		cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x03, scale: 0x00).format()
+	}
+	if (device?.currentValue("humidity") == null) {
+		cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x05, scale: 0x00).format()
+	}
+	if (canReportBattery() || device?.currentValue("battery") == null) {
+		cmds << zwave.batteryV1.batteryGet().format()
+	}
+	
+	getConfigurationNumber().each { idx ->
+		if (state."${getConfigurationInfo(idx, "name")}Refresh" == true) {
+			switch (getConfigurationInfo(idx, "type")) {
+				case "enum":
+					optionValue = getConfigurationInfo(idx, "enumMap").find { it.key == state."${getConfigurationInfo(idx, "name")}" }?.value
+					value = integer2Array(optionValue, getConfigurationInfo(idx, "size"))
+					break
+				case "number":
+					value = integer2Array(state."${getConfigurationInfo(idx, "name")}", getConfigurationInfo(idx, "size"))
+					break
+			}
+			if (value != null) {
+				cmds << zwave.configurationV2.configurationSet(parameterNumber: idx, defaultValue: false, size: getConfigurationInfo(idx, "size"), configurationValue: value).format()
+				cmds << zwave.configurationV2.configurationGet(parameterNumber: idx).format()
+				if (idx == 1) {
+					cmds << zwave.sensorMultilevelV5.sensorMultilevelGet(sensorType: 0x01, scale: optionValue?:0x00).format()
+				}
+				value = null
+			}
+		}
+	}
+	
+	if (state."${getWakeUpInfo("name")}Refresh" == true) {
+		 cmds << zwave.wakeUpV2.wakeUpIntervalSet(nodeid: zwaveHubNodeId, seconds: hour2Second(state.wakeUpInterval)).format()
+		 cmds << zwave.wakeUpV2.wakeUpIntervalGet().format()
+	}
+	
+	sendEvent(name: "checkInterval", value: (hour2Second(state.wakeUpInterval) + 2 * 60) * 2, displayed: false, data: [protocol: "zwave", hubHardwareId: device.hub.hardwareID])
+	
+	return cmds ? delayBetween(cmds, 500) : []
+}
+
+def configurationUpdate(config) {
+	getConfigurationNumber().each { idx ->
+		if (config?."${getConfigurationInfo(idx, "name")}" != null) {
+			if (state."${getConfigurationInfo(idx, "name")}" != config?."${getConfigurationInfo(idx, "name")}") {
+				state."${getConfigurationInfo(idx, "name")}" = config?."${getConfigurationInfo(idx, "name")}"
+				state."${getConfigurationInfo(idx, "name")}Refresh" = true
+			}
+		}
+	}
+}
+
+def configurationCheck(config) {
+	getConfigurationNumber().each { idx ->
+		if (config?."${getConfigurationInfo(idx, "name")}" != null) {
+			if (config?."${getConfigurationInfo(idx, "name")}" == state."${getConfigurationInfo(idx, "name")}") {
+				if (state."${getConfigurationInfo(idx, "name")}Refresh" == true) {
+					state."${getConfigurationInfo(idx, "name")}Refresh" = false
+				}
+			}
+		}
+	}
+}
+
+def getConfigurationNumber() {
+	return [1, 2, 3, 4, 5, 6, 7, 8]
+}
+
+def getConfigurationInfo(num, text) {
+	def parameter = [:]
+	
+	parameter.parameter1name = "TemperatureUnit"
```

Comment:
wouldn't it make more sense to do all of this as a map of maps? Would certainly make iteration etc a lot simpler for you.

i.e.: 

```
def parameter = [:]
parameter[1] = {name: "TemperatureUnit", "title": "Temperature Unit [C/F]" ..}
```

Reply:
I think not. Using a map of maps does not change the iteration way in my case, since I still use the index to make iteration to get key and value. In my case, I use the parameter name as a part of key name of state (i.e. state."${getConfigurationInfo(idx, "name")}Refresh" and state."${getConfigurationInfo(idx, "name")}") to store which configuration is changed by user and check which configuration command needs to be sent.
I think a map of maps that maybe looks better but here it does not make iteration more simpler for me.
I paste the snippet code of this function using a map of maps.
```
def getConfigurationInfo(num, text) {
	def parameter = [:]
	parameter[1] = [
		name: "TemperatureUnit", 
		title: "Temperature Unit [°C/°F]",
		description: "", 
		type: "enum", 
		options: ["°C", "°F"], 
		enumMap: ["°C": 0, "°F": 1], 
		default: "°C", 
		size: 1
	]
	...
	parameter[8] = [
		name: "RetryTimes", 
		title: "Motion notification retry times", 
		description: "",
		type: "number", 
		options: "0..10", 
		enumMap: [], 
		default: 3, 
		size: 1
	]
	return parameter[num.toInteger()]."${text}"
}
```